src/LibDDRV.sol

```diff
@@ -241,52 +189,68 @@ library LibDDRV {
             insert_element(index, newRange);
             newRange.weight += newWeight;
 
-            // enqueue the new range for an update
-            assembly {
-                // Check if the bit j is set in the header
-                if gt(shr(255, shl(sub(255, j), mload(ptr))), 0) {
-                    // If it's not, add the range to the queue
-                    // Set the bit j
-                    mstore(ptr, or(shl(j, 1), mload(ptr)))
-                    // Store the range in the queue
-                    mstore(tail, newRange.slot)
-                    // Update the tail of the queue
-                    tail := add(tail, word)
-                }
-                // Cap off the level queue by incrementing the free memory pointer
-                mstore(fp, add(tail, word))
-            }
-
+            enqueue_range(ptr, head, tail, k, newRange);
         } else {
             // set the new weight of the element
             currentRange.weight += newWeight;
             currentRange.weight -= oldWeight;
         }
 
         // enqueue the current range for an update
+        enqueue_range(ptr, head, tail, j, currentRange);
+
+        update_levels(ptr, head, tail, forest);
+    }
+
+    // Generate a discretely weighted random variate given a uniform random
+    // and a preprocessed forest of trees.
+    function generate(mapping(uint256 => mapping(uint256 => Range)) storage forest, uint256 urv) external {}
+
+    // TODO
+    function get_element(uint256 index) internal returns (Element storage) {revert();}
+
+    function new_queue(bytes32 head) internal returns (bytes32 ptr, bytes32 head1, bytes32 tail) {
+        // Set up an in memory queue object
+        // Qₗ = ∅ OR Qₗ₊₁ = ∅
+        assembly {
+            // Set the queue to the free pointer
+            ptr := mload(fp)
+            // One word is reserved here to act as a header for the queue,
+            // to check if a range is already in the queue.
+            head1 := add(ptr, word)
+            // if head is not supplied (Qₗ = ∅), set tail == head1; else, head
+            tail := head
+            if eq(head, 0) {
```

Comment:
This seems strange. This should just create a new queue, doesn't really matter which level the queue is for.